### PR TITLE
Fix/approximate data compatibility

### PR DESCRIPTION
**Pull request by tc85324**
### PR Type

- Bugfix
- Feature

### Description
Since the introduction of `kernel.compute_mean`, the `gramian_row_mean` method is expected to accept `coreax.data.Data` objects. Compatibility is provided by making `coreax.data.Data` a proper JAX  `ArrayLike` type. This allows `Data.data` to be accessed via `jnp.asarray(Data(...))`.
    
In addition, the new `kernel.compute_mean` method allows some duplicated code to be removed from the old approximator implementations.

To reduce duplication, an `as_data` method has been added, and the `kernel.gramian_row_mean` docstring has been simplified (including typehints for block_size).

### How Has This Been Tested?
All tests pass

### Does this PR introduce a breaking change?
No

### Checklist before requesting a review
- [ ] I have made sure that my PR is not a duplicate.
- [ ] My code follows the style guidelines of this project.
- [ ] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have performed a self-review of my code.
- [ ] I have made corresponding changes to the documentation.
- [ ] My changes generate no new warnings.
- [ ] I have added tests that prove my fix is effective or that my feature works.
- [ ] New and existing unit tests pass locally with my changes.
- [ ] Any dependent changes have been merged and published in downstream modules.

### CODE REVIEW

File: coreax/approximation.py [hunk 1]
```diff
@@ -180,7 +183,7 @@ def gramian_row_mean(self, x: ArrayLike, **kwargs) -> Array:
         :return: Approximation of the base kernel's Gramian row-mean
         """
         del kwargs
```
Comment (pc532627) [inline coreax/approximation.py:185]: Just as a sense check, do you know when this was introduced, and what it needs doing at the top of all of these methods? (deleting the `kwards`)?

File: coreax/approximation.py [hunk 2]
```diff
@@ -180,7 +183,7 @@ def gramian_row_mean(self, x: ArrayLike, **kwargs) -> Array:
...
-        data = jnp.atleast_2d(x)
+        data = jnp.asarray(x)
```
Comment (pc532627) [inline coreax/approximation.py:186]: In all previous cases, we were assuming data was handled as a 2D JAX array (with `jnp.atleast_2d`). I assume `jnp.asarray` is needed for compatibility with `x` if `x` is of type `Data`? Can we not keep the assumption that everything is worked with as a 2D array?

File: coreax/data.py [hunk 3]
```diff
@@ -207,19 +207,23 @@ class Data(eqx.Module):
...
     def __init__(
         self,
-        data: Shaped[ArrayLike, " n d"],
+        data: Shaped[ArrayLike, " n *d"],
         weights: Shaped[ArrayLike, " n"] | None = None,
     ):
         """Initialise Data class."""
-        self.data = jnp.atleast_2d(data)
-        n = self.data.shape[0]
+        self.data = jnp.asarray(data)
```
Comment (pc532627) [inline coreax/data.py:219]: Same comment about `jnp.asarray` vs `jnp.atleast_2d` - is this needed and has the code been tested if 1d arrays are passed?

**Review by pc532627 — CHANGES_REQUESTED** [review]

File: coreax/approximation.py (hunk 1 above)
Comment (tc85324) [reply coreax/approximation.py:185]: The kwargs parameter is for Liskov, the del is to supress pylint's complaint of an unused variable.

File: coreax/approximation.py (hunk 2 above)
Comment (tc85324) [reply coreax/approximation.py:186]: We only test for 2D at the moment. I don't like the usage of `atleast_2d` as it is not explicit to the user where this dimension will be added (the leading axis). If the user passes an n-vector (believing this represents n scalar observations) then `atleast_2d` will silently give them the wrong result. This is because `atleast_2d` converts the vector to an array with shape `(1, n)`, rather than `(n, 1)`. 

The difficulty is that different users may assume different behaviour for vectors. Some may (correctly in this case) assume an n-vector represents a single n-dimensional observation, whereas others may (incorrectly in this case) assume an n-vector represents n 1-dimensional observations.

Because of this, I think we only have three options: 
1. throw an error when passed an array that is not 2D `if jnp.ndim(x) != 2: raise ValueError("Must be 2D")`
2. allow k-dimensional arrays and let JAX handle any resulting incompatible mathematical operations.
3. use beartype and jaxtyping to provide (optional) runtime type checking of the array dimensions.

If we start using jaxtyping everywhere then the expectations on user data become clearer before runtime (at programming time).

Addressing this issue properly across the entire codebase is out of scope for this pull, so I would prefer to keep this as is, or just wrap the `asarrays` in an `atleast_2d` just to preserve the previous behaviour (whichever you would prefer).

Comment (pc532627) [reply coreax/approximation.py:186]: @tc85324 Thanks for highlighting the options. I agree that it is not explicit to the user what will happen with particular inputs and could lead to confusion. My proposal is to wrap the `asarrays` in an `atleast_2d` call to preserve the previous behavior for now, but I will also create a ticket that details this discussion and in resolving that, we do one of the three listed options and remove the wrapped `atleast_2d` calls. How does that sound?

Comment (pc532627) [reply coreax/approximation.py]: For reference this is now discussed in issue #630

File: coreax/data.py (hunk 3 above)
Comment (pc532627) [reply coreax/data.py:219]: @tc85324 this needs a `atleast_2d` wrapper I believe

**Review by pc532627 — CHANGES_REQUESTED** [review]
I think only 1 line has a change required now

File: coreax/data.py (hunk 3 above)
Comment (tc85324) [reply coreax/data.py:219]: The type hint indicates that d is optional, so will be treated as an n-vector. In this case we have logical consistency as weights is expected to be an n-vector and the docstrings make clear that each weight is associated with a leading index of the data. If you have atleast_2d here, you end up with the confusing scenario where you pass an `n-vector` of data but are only expected to pass a single scalar weight. In addition, there is no reason why data should have to be 2d here, whereas at the point of use (in the gramian_row_mean of the approximators) the mathematical operations doe require 2d.

**Review by pc532627 — APPROVED** [review]
